shared/src/main/java/com/couchbase/lite/AbstractReplicator.java

```diff
@@ -629,19 +635,38 @@ public void run() {
         c4ReplListener.statusChanged(c4repl, c4ReplStatus, this);
     }
 
-    private void documentEnded(boolean pushing, String docID, String revID, int flags, C4Error error, boolean trans) {
-        DocumentReplication update = new DocumentReplication((Replicator) this,
-                flags == C4Constants.C4RevisionFlags.kRevDeleted,
-                 pushing, docID, revID, flags, error, trans);
-        if (!pushing && error.getDomain() == LiteCoreDomain && error.getCode() == kC4ErrorConflict) {
-            // Conflict pulling a document -- the revision was added but app needs to resolve it:
-            Log.i(TAG, "%s: pulled conflicting version of '%s'", this, docID);
+    private void documentEnded(boolean pushing, C4DocumentEnded[] documents) {
+        List<ReplicatedDocument> docs = new ArrayList<ReplicatedDocument>();
+        for (C4DocumentEnded c4document: documents) {
+            boolean isAccessRemoved = false;
+            boolean isDeleted = false;
+            C4Error error = c4document.getC4Error();;
             try {
-                this.config.getDatabase().resolveConflictInDocument(docID);
-            } catch (CouchbaseLiteException ex) {
-                Log.e(TAG, "Failed to resolveConflict: docID -> %s", ex, docID);
+                isAccessRemoved = config.getDatabase().c4db.get(c4document.getDocID(), false).accessRemoved();
+                isDeleted = config.getDatabase().c4db.get(c4document.getDocID(), false).deleted();
```

Comment:
There is no need to call c4db to get the doc. The information is already in the c4document.flag.

Reply:
I think documentEnded() method can be implemented as this:

```Java
    private void documentEnded(boolean pushing, C4DocumentEnded[] docEnds) {
        List<ReplicatedDocument> docs = new ArrayList<ReplicatedDocument>();
        
        for (C4DocumentEnded docEnd : docEnds) {
            String docID = docEnd.getDocID();
            boolean isDeleted = (docEnd.getFlags() & C4Constants.C4RevisionFlags.kRevDeleted) != 0;
            boolean isAccessRemoved = (docEnd.getFlags() & C4Constants.C4RevisionFlags.kRevPurged) != 0;
            C4Error error = docEnd.getC4Error();
            if (!pushing && docEnd.getErrorDomain() == LiteCoreDomain && docEnd.getErrorCode() == kC4ErrorConflict) {
                // Conflict pulling a document -- the revision was added but app needs to resolve it:
                Log.i(TAG, "%s: pulled conflicting version of '%s'", this, docID);
                try {
                    this.config.getDatabase().resolveConflictInDocument(docID);
                    error = new C4Error();
                } catch (CouchbaseLiteException ex) {
                    Log.e(TAG, "Failed to resolveConflict: docID -> %s", ex, docID);
                }
            }
            ReplicatedDocument doc = new ReplicatedDocument(isDeleted, isAccessRemoved, docID,
                    docEnd.getRevID(), error, docEnd.errorIsTransient());
            docs.add(doc);
        }

        DocumentReplication replication = new DocumentReplication((Replicator) this, pushing, docs);
        notifyDocumentEnded(replication);
    }
```